File: rscuad_camera/gateway/gateway.cpp
#include "gateway.h"

#include <unistd.h>     // sleep
#include <string>
#include <iostream>
#include <mutex>
#include <thread>

#define TIME_DELAY 50000

std::mutex mtxData;
std::mutex mtxImage;


int 				Gateway::m_sockfdData;
int 				Gateway::m_sockfdImage;
struct sockaddr_in 	Gateway::m_addrData;
struct sockaddr_in 	Gateway::m_addrImage;
std::vector<u_char>	Gateway::m_frame;

struct Gateway::DataResponse *Gateway::s_dataResponse;


Gateway::Gateway()
: m_portData(9131)
, m_portImage(9132)
, m_length(10)
{
	s_dataResponse = new DataResponse;
}

Gateway::~Gateway()
{
}

void Gateway::Init()
{
	if ( (m_sockfdData = socket(AF_INET, SOCK_DGRAM, 0)) < 0 ) 
	{
		perror("socket creation failed");
		exit(EXIT_FAILURE);
	}
	

	memset(&m_addrData, 0, sizeof(m_addrData));
	m_addrData.sin_family = AF_INET;
	m_addrData.sin_port = htons(m_portData);
	m_addrData.sin_addr.s_addr = inet_addr("127.0.0.1");

	if ( (m_sockfdImage = socket(AF_INET, SOCK_DGRAM, 0)) < 0 ) 
	{
		perror("socket creation failed");
		exit(EXIT_FAILURE);
	}

	memset(&m_addrImage, 0, sizeof(m_addrImage));
	m_addrImage.sin_family = AF_INET;
	m_addrImage.sin_port = htons(m_portImage);
	m_addrImage.sin_addr.s_addr = inet_addr("127.0.0.1");

	Gateway::CreateAsync();
}

void Gateway::CreateAsync()
{
	std::thread workerSendData(AsyncData);
	workerSendData.detach();

	std::thread workerSendImage(AsyncImage);
	workerSendImage.detach();

}

void Gateway::AsyncData()
{
	for(;;)
	{
		std::lock_guard<std::mutex> guard(mtxData);

		std::string data = s_dataResponse->isBall +" "+ s_dataResponse->ballX +" "+ s_dataResponse->ballY +" "+
							s_dataResponse->robotX +" "+ s_dataResponse->robotY +" "+ s_dataResponse->point +" "+ 
							s_dataResponse->isTarget +" "+ s_dataResponse->targetX +" "+ s_dataResponse->targetY;

		sendto(m_sockfdData, data.c_str(), data.length(), MSG_CONFIRM, 
						(const struct sockaddr *) &m_addrData, sizeof(m_addrData));

		usleep(TIME_DELAY);

	}
}

void Gateway::AsyncImage()
{
	for(;;)
	{
		std::lock_guard<std::mutex> guard(mtxImage);

		sendto(m_sockfdImage, m_frame.data(), m_frame.size(), MSG_CONFIRM, 
					(const struct sockaddr *) &m_addrImage, sizeof(m_addrImage));

		usleep(TIME_DELAY);
	}
}

void Gateway::Update(const std::vector<u_char>& frame)
{
	m_frame.assign(frame.begin(), frame.end());
}

void Gateway::Update(bool isBall, int ballX, int ballY, 
						int robotX, int robotY, int point,
						bool isTarget, int targetX, int targetY)
{
	s_dataResponse->isBall = FloatToString((int)isBall);
	s_dataResponse->ballX = FloatToString(ballX);
	s_dataResponse->ballY = FloatToString(ballY);
	s_dataResponse->robotX = FloatToString(robotX);
	s_dataResponse->robotY = FloatToString(robotY);
	s_dataResponse->point = FloatToString(point);
	s_dataResponse->isTarget = FloatToString((int)isTarget);
	s_dataResponse->targetX = FloatToString(targetX);
	s_dataResponse->targetY = FloatToString(targetY);
}

std::string Gateway::FloatToString(const float data) const
{
    std::stringstream strs;
    strs << data;
    return strs.str();
}

File: rscuad_camera/gateway/gateway.py
# 
# 	Title	: AI Gateway
# 
# 	Author	: Danu Andrean
# 	Create	: 2023
# 	Update 	: 2024
# 	
# 	Note	: gateway for data python
# 

USE_ROS = True

import socket
import sys
import numpy as np
import cv2 as cv
import threading
import time

if USE_ROS:
	import rospy
	from std_msgs.msg import String
	from rscuad.msg import DataCamera

	import rospy 
	from sensor_msgs.msg import Image 
	from cv_bridge import CvBridge 


class Gateway:
	
	ip = "127.0.0.1"
	port_data = 9131
	port_image = 9132
	data = "0 0 0 0 0"
	frame = 0

	rospy.init_node('camera_module', anonymous=True)
	pub_image = rospy.Publisher('video_camera', Image, queue_size=10)
	pub_data = rospy.Publisher('data_camera', DataCamera, queue_size=10)

	rate = rospy.Rate(10) # 10hz
	br = CvBridge()

	def __init__(self):
		self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

	def init(self):
		data = threading.Thread(target=Gateway.async_data,  args = (self,), daemon=True)
		image = threading.Thread(target=Gateway.async_image, args = (self,), daemon=True)
 
		data.start()
		image.start()
		print("Running in {}".format(self.ip))

	def async_data(self):
		while(1):
			self.sock.sendto(self.data.encode('utf-8'), (self.ip, self.port_data))
			time.sleep(0.1)
	def async_image(self):
		while(1):
			_, encoded_frame = cv.imencode(".jpg", self.frame, [int(cv.IMWRITE_JPEG_QUALITY), 70])
			self.sock.sendto(encoded_frame.tobytes(), (self.ip, self.port_image))
			time.sleep(0.1)

	def update_data(self, is_ball, ball_x, ball_y, goal_x=0, goal_y=0, is_goal=False, is_robot=False, robot_x=0, robot_y=0):
		if USE_ROS:
			dc = DataCamera()
			dc.ball = is_ball
			dc.ball_x = ball_x
			dc.ball_y = ball_y
			dc.point = is_goal
			dc.loc_x = goal_x
			dc.loc_y = goal_y
			dc.target = is_robot
			dc.target_x = robot_x
			dc.target_y = robot_y
			
			self.pub_data.publish(dc)
		else:
			self.data = str(is_ball)+" "+str(ball_x)+" "+str(ball_y)+" "+str(loc_x)+" "+str(loc_y)+" "+str(point)+" "+ str(is_target)+" "+str(target_x)+" "+str(target_y)

	def update_image(self, frame):
		if USE_ROS:
			self.pub_image.publish(self.br.cv2_to_imgmsg(frame))
		else:
			self.frame = frame

	def get_ros(self):
		return USE_ROS
		
